### DISCUSSION TITLE
Update project removes limit-modules from .classpath

### DISCUSSION THREAD

**Discussion by Vertganti** [eclipse-m2e/m2e-core#1769] [Q&A]
We are currently migrating a large project from Java 8 to Java 21. The project uses facets and has the Java facet enabled for Version 21. We use the `javax.smartcardio` package, which was moved to the module [`java.smartcardio`](https://docs.oracle.com/en/java/javase/21/docs/api/java.smartcardio/module-summary.html) in Java 9. Unfortunately the module is not included in the default module dependencies. It has to be added by opening the project configuration in Eclipse, going to `Java Build Path` -> `Module Dependencies` -> `Add System Module...`, selecting the `java.smartcardio` entry, clicking `Add` and then `Apply and Close` - see [this StackOverflow answer](https://stackoverflow.com/a/58951434/8254045). This will add the `module` and `limit-modules` attribute to one of the `classpathentry` entries of the `.classpath` file of the project:

```
...
	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-21">
		<attributes>
			<attribute name="module" value="true"/>
			<attribute name="limit-modules" value="java.se,jdk.accessibility,jdk.dynalink,jdk.httpserver,jdk.incubator.vector,jdk.jartool,jdk.javadoc,jdk.jconsole,jdk.jshell,jdk.jsobject,jdk.management.jfr,jdk.net,jdk.nio.mapmode,jdk.sctp,jdk.security.auth,jdk.security.jgss,jdk.unsupported,jdk.unsupported.desktop,jdk.xml.dom,java.smartcardio"/>
			<attribute name="maven.pomderived" value="true"/>
		</attributes>
	</classpathentry>
...
```

Unfortunately those two lines are deleted when we run `Maven` -> `Update Project...` and we get stuck in an infinite build loop. We have tried adding `--add-modules` and `--limit-modules` args to the `maven-compiler-plugin` build step in our `pom.xml` as below:

```
...
	<plugin>
		<groupId>org.apache.maven.plugins</groupId>
      		<artifactId>maven-compiler-plugin</artifactId>
      		<version>3.13.0</version>
      		<configuration>
			<encoding>ISO-8859-1</encoding>      	
      			<release>21</release>
      			<compilerArgs>
	            		<arg>--add-modules</arg>
	            		<arg>java.smartcardio</arg>
                                <arg>--limit-modules</arg>
	            		<arg>java.se,jdk.accessibility,jdk.dynalink,jdk.httpserver,jdk.incubator.vector,jdk.jartool,jdk.javadoc,jdk.jconsole,jdk.jshell,jdk.jsobject,jdk.management.jfr,jdk.net,jdk.nio.mapmode,jdk.sctp,jdk.security.auth,jdk.security.jgss,jdk.unsupported,jdk.unsupported.desktop,jdk.xml.dom,java.smartcardio</arg>
	        	</compilerArgs>
      		</configuration>
	</plugin>
...
```

We tried both args separately and both together but the lines are always removed from the `.classpath`. We have been unable to find any documentation on how to add a module dependency through the `pom.xml` and hoped that the compiler arguments would do it. Is there a configuration option for this elsewhere in the `pom.xml` or the compiler plugin? Or is this a bug/missing feature?

**Comment by roeltje25** [comment]
I have this problem as well. module settings should reflect pom specifications. Also in non modular projects.

**Reply by roeltje25** [reply]
It seems that this is explicitly not supported at the moment as I look at the code here:
https://github.com/CsCherrYY/m2e-core/blob/64c348c58f1eebd2254c25d5346a4288c79227c7/org.eclipse.m2e.jdt/src/org/eclipse/m2e/jdt/internal/ModuleSupport.java#L609

Is there anyone able and willing to fix this?
